tsl/src/continuous_aggs/create.c

```diff
@@ -1731,3 +1750,363 @@ tsl_process_continuous_agg_viewstmt(ViewStmt *stmt, const char *query_string, vo
 	cagg_create(stmt, query, &timebucket_exprinfo, with_clause_options);
 	return true;
 }
+
+/*
+ * update the view definition of an existing continuous aggregate
+ */
+void
+cagg_update_view_definition(ContinuousAgg *agg, Hypertable *mat_ht,
+							WithClauseResult *with_clause_options)
+{
+	ListCell *lc1, *lc2;
+	int sec_ctx;
+	Oid uid, saved_uid;
+	/* cagg view created by the user */
+	Oid user_view_oid = relation_oid(agg->data.user_view_schema, agg->data.user_view_name);
+	Relation user_view_rel = relation_open(user_view_oid, AccessShareLock);
+	Query *user_query = get_view_query(user_view_rel);
+	relation_close(user_view_rel, AccessShareLock);
+
+	Oid direct_view_oid = relation_oid(agg->data.direct_view_schema, agg->data.direct_view_name);
+
+	Relation direct_view_rel = relation_open(direct_view_oid, AccessShareLock);
+	Query *direct_query = get_view_query(direct_view_rel);
+
+	CAggTimebucketInfo timebucket_exprinfo = cagg_validate_query(direct_query);
+
+	relation_close(direct_view_rel, AccessShareLock);
+	FinalizeQueryInfo fqi;
+	MatTableColumnInfo mattblinfo;
+	ObjectAddress mataddress = {
+		.classId = RelationRelationId,
+		.objectId = mat_ht->main_table_relid,
+	};
+
+	mattablecolumninfo_init(&mattblinfo, NIL, NIL, copyObject(direct_query->groupClause));
+	finalizequery_init(&fqi, direct_query, &mattblinfo);
+
+	Query *view_query = finalizequery_get_select_query(&fqi, mattblinfo.matcollist, &mataddress);
+
+	/* adjust varnos in targetlist */
+	Assert(list_length(view_query->rtable) > 1);
+	foreach (lc1, view_query->targetList)
+	{
+		ChangeVarNodes((Node *) lfirst_node(TargetEntry, lc1),
+					   1,
+					   list_length(view_query->rtable),
+					   0);
+	}
+
+	if (with_clause_options[ContinuousViewOptionMaterializedOnly].parsed == BoolGetDatum(false))
+		view_query = build_union_query(&timebucket_exprinfo, &mattblinfo, view_query, direct_query);
+
+	/*
+	 * adjust names in the targetlist of the updated view to match the view definition
+	 */
+	Assert(list_length(view_query->targetList) == list_length(user_query->targetList));
+
+	forboth (lc1, view_query->targetList, lc2, user_query->targetList)
+	{
+		TargetEntry *view_tle, *user_tle;
+		view_tle = lfirst_node(TargetEntry, lc1);
+		user_tle = lfirst_node(TargetEntry, lc2);
+		view_tle->resname = user_tle->resname;
+	}
+
+	SWITCH_TO_TS_USER(NameStr(agg->data.user_view_schema), uid, saved_uid, sec_ctx);
```

Comment:
need an exclusive lock on view i.e. LockRelationOid(...) before changing the definition. The lock should only be released at the end of the command and not by this function. please add a comment noting that as well.

Reply:
We should also add a concurrent test (isolation tests for cont. aggregates) to test replacing view waits for other txns accessing the view
1. Lock the view in share mode.
Session 2 . Refresh mat view.
Session 3.  replace view definition
session 2 and 3 wait
Session 1 ROLLBACK
Session 2 and Session 3 should complete.

Reply:
postgres already aquires the required locks

Reply:
@svenklemm Could you elaborate on this? Don't we have to acquire the locks explicitly on the view since we intercept the create/replace view code path?